### PR TITLE
update: Start work on results reproduction

### PR DESCRIPTION
**Pull request by mrshu**
- Start work on reproduction of results published in
  http://dai.fmph.uniba.sk/projects/smelyzajko/Michal_Moravcik_DP/Michal_Moravcik_Smely_Zajko.pdf

Signed-off-by: mr.Shu mr@shu.io

### CODE REVIEW

File: model_mlp.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+def prepare_dataset(dataset, window_x=5, window_y=5):
+    Xs = []
+    ys = []
+    for datum in dataset:
+        img = datum['img']
+        mask = datum['mask']
+        for selection, mask_selection in prepare_image(img, mask):
+            Xs.append(selection)
+            ys.append(mask_selection)
+    return np.array(Xs), np.array(ys)
```
Comment (vidriduch) [inline model_mlp.py:27]: @mrshu Shouldn't be training data divided further to training and validation data. And then evaluate final score once on test data?

File: model_mlp.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+dataset = utils.dataset_from_folder('./plzen/test')
+X_test, y_test = prepare_dataset(dataset)
+
+model = Sequential()
+model.add(Dense(8, input_shape=(75,)))
+model.add(Activation('sigmoid'))
+model.add(Dense(1))
+model.add(Activation('sigmoid'))
+
+sgd = SGD()
```
Comment (vidriduch) [inline model_mlp.py]: @mrshu As far as I understood the paper, they tried iProp which based on their description should be something like RMSprop or SGD with momentum. Also based on my experiments if we going to use RMSprop best learning rate was around 0.001-0.01

**Comment by vidriduch** [comment]
@mrshu Except those two small comments this looks good. It is very clear and pretty. :+1:

**Comment by mrshu** [comment]
@Foxiy thank you for your feedback, I tried both things you pointed out by:
- Using the `validation_split` parameter of the `fit` function which will split the training data into a "training" set and a "validation" set using the specified ratio
- Using `RMSProp` with the learning rate set to `0.01`.

I believe this is ready for another pass.

File: model_mlp.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+rms = RMSprop(lr=0.01)
+model.compile(loss='mse', optimizer=rms)
+
+model.fit(X_train, y_train,
+          batch_size=batch_size,
+          nb_epoch=nb_epoch, verbose=1,
+          validation_split=0.1)
+
+score = model.evaluate(X_test, y_test)
+print('Test MSE: {}'.format(score))
```
Comment (vidriduch) [inline model_mlp.py:72]: @mrshu Lastly, this on the same line as output of `model.evaluate(X_test, y_test)` so maybe add `\n`

File: model_mlp.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+model = Sequential()
+model.add(Dense(8, input_shape=(75,)))
+model.add(Activation('sigmoid'))
+model.add(Dense(1))
+model.add(Activation('sigmoid'))
+
+rms = RMSprop(lr=0.01)
+model.compile(loss='mse', optimizer=rms)
+
+model.fit(X_train, y_train,
```
Comment (vidriduch) [inline model_mlp.py]: @mrshu The fit function actually returns history objects so we can visualize what is going on be just doing

``` python
history = model.fit(X_train, y_train,
                    batch_size=batch_size, 
                    nb_epoch=nb_epoch, verbose=1,
                    validation_split=0.1)

plt.plot(history.history['loss'], 'o', label='loss')   
plt.plot(history.history['val_loss'], '-go', label='validation loss')
plt.xlabel('epoch')
plt.ylabel('loss')
plt.legend()
plt.show()
```

of course we need to import `pyplot` from `matplotlib`

**Comment by vidriduch** [comment]
@mrshu Otherwise, this looks good. I don't think we are going get better results without help of the author of original paper since there are no details about training included in his work.  Either way we came pretty close and I think with more epoch we would get published results. This can be merged if you don't feel like adding those little formatting things or right after

**Comment by mrshu** [comment]
@Foxiy thanks for the comments again, I'd like you to take another look and then merge it in.

File: model_mlp.py (hunk 1 above)
Comment (mrshu) [inline model_mlp.py:27]: @Foxiy I tried to counter this a bit later down by using `validation_split`.

File: model_mlp.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                    validation_split=0.1)
+
+plt.plot(history.history['loss'], 'o', label='loss')
+plt.plot(history.history['val_loss'], '-go', label='validation loss')
+plt.xlabel('epoch')
+plt.ylabel('loss')
+plt.legend()
+plt.show()
+
+score = model.evaluate(X_test, y_test, verbose=0)
```
Comment (mrshu) [inline model_mlp.py:71]: @Foxiy This line should no longer print out anything. I am not sure if the `\n` is still necessary then.

**Comment by vidriduch** [comment]
@mrshu Thanks for the formatting :+1: :)
